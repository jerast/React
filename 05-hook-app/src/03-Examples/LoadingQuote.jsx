import eclipse from '../assets/eclipse.svg'

export const LoadingQuote = () => {
   return <img src={ eclipse } alt="...Loading" />;
};
