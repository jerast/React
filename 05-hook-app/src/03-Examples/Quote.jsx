import PropTypes from 'prop-types';
import { useLayoutEffect, useRef } from 'react';

export const Quote = ({ quote, author }) => {

	const quoteRef = useRef();

	useLayoutEffect(() => {
		console.log( quoteRef.current.getBoundingClientRect().height )
	}, [])

	return (
		<blockquote className="blockquote">
			<p ref={ quoteRef }>"{ quote }"</p>
			<span>- { author } -</span>
		</blockquote>
	);
};

Quote.propTypes = {
   quote: PropTypes.string.isRequired,
   author: PropTypes.string.isRequired,
};
