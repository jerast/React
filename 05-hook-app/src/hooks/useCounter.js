import { useState } from "react";

export const useCounter = ( initialValue = 10 ) => {

	const [counter, setCounter] = useState( initialValue );

	const increment = ( steps = 1 ) => setCounter( counter => counter + steps );
	const decrement = ( steps = 1 ) => setCounter( counter => counter - steps );
	const reset = () => setCounter( initialValue );

	return {
		counter,
		increment,
		decrement,
		reset
	};
};