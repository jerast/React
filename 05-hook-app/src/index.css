@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
	* {
		@apply box-border;
	}

	body {
		@apply bg-zinc-800;
	}
	#root {
		@apply w-full grid gap-3 justify-items-center p-5  text-white;
	}

	.title {
		@apply text-center text-4xl my-4;
	}
	.subtitle {
		@apply text-center text-2xl my-4;
	}

	button {
		@apply px-5 py-2 bg-zinc-500 text-white rounded-full disabled:opacity-50 disabled:cursor-default;
	}
	input {
		@apply bg-zinc-500 px-5 py-2 rounded-xl w-full;
	}

	blockquote {
		@apply w-full max-w-[500px];
	}
	blockquote p {
		@apply font-[cursive] text-6xl m-3 text-center;
	}
	blockquote span {
		@apply text-end w-full inline-block font-light font-mono italic text-sm;
	}
}
