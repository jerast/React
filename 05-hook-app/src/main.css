@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
	* {
		@apply box-border select-none;
	}

	body {
		@apply bg-zinc-800;
	}
	#root {
		@apply w-full grid gap-3 justify-items-center p-5  text-white;
	}

	.title {
		@apply text-center text-4xl;
	}
	.subtitle {
		@apply text-center text-xl text-zinc-400;
	}

	button {
		@apply px-5 py-2 bg-zinc-700 text-white rounded-full disabled:opacity-50 disabled:cursor-default outline-none ease-in-out duration-75 hover:bg-zinc-800;
	}
	input {
		@apply bg-zinc-600 px-5 py-2 rounded-xl w-full;
	}
	form {
		@apply flex rounded-lg overflow-hidden shadow-md my-8
	}
	form > * {
		@apply rounded-none outline-none 
	}

	ul {
		@apply grid gap-2 w-full max-w-[500px]
	}
	.list-item {
		@apply w-full relative h-12 cursor-pointer ease-in-out duration-75
	}
	.list-item span {
		@apply w-full h-full rounded-full inline-flex items-center bg-zinc-500  hover:bg-zinc-600 ease-in-out duration-75 pl-5 shadow-md;
	}
	.list-item button {
		@apply text-2xl absolute right-0 px-5
	}
	.list-item.done {
		@apply  opacity-30;
	}
	.list-item.done span {
		@apply bg-zinc-600 line-through shadow-none;
	}

	blockquote {
		@apply w-full max-w-[500px];
	}
	blockquote p {
		@apply font-[cursive] text-6xl m-3 text-center;
	}
	blockquote span {
		@apply text-end w-full inline-block font-light font-mono italic text-sm;
	}
}
