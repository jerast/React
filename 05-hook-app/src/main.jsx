import React from 'react';
import ReactDOM from 'react-dom/client';
import { Padre } from './07-tarea-memo/Padre';
import './index.css';

// import { HooksApp } from './HooksApp'
// import { CounterApp } from './01-State/CounterApp'
// import { CounterWithCustomHook } from './01-State/CounterWithCustomHook';
// import { FormWithCustomHook } from './02-Effect/FormWithCustomHook';
// import { SimpleForm } from './02-Effect/SimpleForm';
// import { MultipleCustomHooks } from './03-Examples/MultipleCustomHooks';
// import { FocusScreen } from './04-Ref/FocusScreen';
// import { Layout } from './05-LayourEffect/Layout';
// import { Memorize } from './06-Memos/Memorize';
// import { MemoHook } from './06-Memos/MemoHook';
// import { CallbackHook } from './06-Memos/CallbackHook';

ReactDOM.createRoot(document.getElementById('root')).render(
	<>
		{/* <React.StrictMode> */}
			{/* <HooksApp /> */}
			{/* <CounterApp /> */}
			{/* <CounterWithCustomHook /> */}
			{/* <SimpleForm /> */}
			{/* <FormWithCustomHook /> */}
			{/* <MultipleCustomHooks /> */}
			{/* <FocusScreen /> */}
			{/* <Layout /> */}
			{/* <Memorize /> */}
			{/* <MemoHook /> */}
			{/* <CallbackHook /> */}
			<Padre />
		{/* </React.StrictMode> */}
	</>
);
