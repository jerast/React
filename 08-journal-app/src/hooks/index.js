
// Hooks
export * from './useForm';
export * from './useCheckAuth';