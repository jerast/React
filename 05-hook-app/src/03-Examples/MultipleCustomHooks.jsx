import { useCounter, useFetch } from '../hooks';
import { LoadingQuote, Quote } from './';

export const MultipleCustomHooks = () => {
	const { counter, increment } = useCounter(1);
	const { data, isLoading	} = useFetch(`https://api.breakingbadquotes.xyz/v1/quotes/${ counter }`);

	const { author, quote } = !isLoading && data[0];

	return <>
		{
			( isLoading ) 
				? <LoadingQuote /> 
				: <Quote quote={ quote } author={ author } />
		}
		<button onClick={ increment } disabled={ isLoading }>Next Quote</button>
	</>
};

// 
// "When you have children, you always have family. 
// They will always be your priority, your responsibility. 
// And a man, a man provides. 
// And he does it even when he's not appreciated or respected or even loved. 
// He simply bears up and he does it. 
// Because he's a man."
// 
// - Gustavo Fring -
// 