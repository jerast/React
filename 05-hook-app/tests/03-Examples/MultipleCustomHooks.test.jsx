import { fireEvent, render, screen } from "@testing-library/react";
import { MultipleCustomHooks } from "../../src/03-Examples/MultipleCustomHooks";
import { useCounter, useFetch } from "../../src/hooks";

jest.mock('../../src/hooks/useFetch');
jest.mock('../../src/hooks/useCounter');

describe('Testing on <MultipleCustomHooks /> component', () => {
   
   const mockIncrement = jest.fn();
   useCounter.mockReturnValue({ 
      counter: 1,
      increment: mockIncrement,
   });

   beforeEach( () => jest.clearAllMocks() );


   test('should match the snapshot', () => {
      useFetch.mockReturnValue({ 
         data: null,
         isLoading: true,
      });
      const { container } = render( <MultipleCustomHooks /> );

      expect( container ).toMatchSnapshot();
   });

   test('should display the component by default', () => {
      useFetch.mockReturnValue({ 
         data: null,
         isLoading: true,
      });
      render( <MultipleCustomHooks /> );

      expect( screen.getByRole('img') ).toBeTruthy();
      expect( screen.getByRole('button', { name: 'Next Quote' }).disabled ).toBeTruthy();
   });

   test('should display a Quote', () => {
      const testingData = [{ author: 'Jerast', quote: 'Hola Mundo' }];
      useFetch.mockReturnValue({ 
         data: testingData,
         isLoading: false,
      });
      render( <MultipleCustomHooks /> );
      
      expect( screen.getByText( testingData[0].author , { exact: false }) ).toBeTruthy();
      expect( screen.getByText( testingData[0].quote, { exact: false } ) ).toBeTruthy();
      expect( screen.getByRole( 'button', { name: 'Next Quote'} ).disabled ).toBeFalsy();
   });

   test('should call the increment function', () => {
      const testingData = [{ author: 'Jerast', quote: 'Hola Mundo' }];
      useFetch.mockReturnValue({ 
         data: testingData,
         isLoading: false,
      });
      render( <MultipleCustomHooks /> );

      const nextButton = screen.getByRole('button', { name: 'Next Quote' });
      fireEvent.click( nextButton );
      
      expect( mockIncrement ).toHaveBeenCalled();
   });

});