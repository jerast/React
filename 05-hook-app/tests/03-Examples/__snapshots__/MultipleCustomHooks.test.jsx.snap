// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on <MultipleCustomHooks /> component should match the snapshot 1`] = `
<div>
  <img
    alt="...Loading"
    src="eclipse.svg"
  />
  <button
    disabled=""
  >
    Next Quote
  </button>
</div>
`;
